doc/developer/design/20230802_webhook_idempotency.md

```diff
@@ -0,0 +1,403 @@
+the data from the upsert envelope.
+
+#### Idempotency guarantees
+
+We plan to offer `UPSERT` idempotency, which means that we will prevent adding
+state to a persist shard if we receive a second write whose `kv` pair matches
+one we have in the `UPSERT` state.
+
+However, this means if we receive multiple requests with the same idempotency
+token, we will retract the initial state and leave the last-seen value in the
+persist shard. Because we plan have add the time the request was received to the
+row of data we append, this means we will continually update the time the
+request was written and extend the lifetime of the idempotency key from the last
+time the write occurred.
+
+I think this is reasonable, but wanted to make sure this behavior's is
+explicitly understood as a tradeoff of storing the time.
+
+**Alternative**
+
+It's possible to achieve a similar outcome by increasing the compaction period
+for the remap shard such that we can get the time at which each event occurred
+in Materialize time--this feedback loop seems challenging to instrument,
+however.
+
+### Architectural overview
+
+#### Dataflow
+
+##### Source rendering
+
+The rendered dataflow will simply read updates out of a channel and give them to
+a data capability at some cadence, downgrading on either some interval's tick or
+when the amount of buffered data reaches some threshold.
+
+Providing users a threshold of requests to buffer offers some knob to control
+throughput vs. latency for the source.
+
+**Timestamping**
+
+The native timestamp of the source could just be a `u64`, which we increment for
+each batch we produce.
+
+The annoying corner of this, though, is that the latency of any request is
+limited by the frequency with which we reclock batches. The most feasible option
+I see here is providing users the option to reclock on a shorter period. The
+current period is 1s, but it could be set to some configurable duration. I'm not
+sure how low this could be set to produce a meaningful effect for the user.
+
+**Request responses**
+
+To provide durability guarantees for webhook sources, we cannot respond to the
+client with a `200` message until we are sure that the request has been
+committed to the shard.
+
+We can retain an in-memory map of `BTreeMap<u64, BTreeSet<IdempotencyKey>>`, and
+then as we get feedback that a batch has been committed can send back all
+idempotency keys committed along response edge.
+
+This has the side-effect of forcing us to send `200` responses on writes whose
+idempotency keys we've already seen because the entire batch commits, i.e. we do
+not have the ability in the rendering to understand if a write committed _now_
+or was already present. I don't believe this to be problematic--a `500` response
+followed by a `200` response, irrespective of the "underlying" result of the
+`500` response still unambiguously states that the write committed.
+
+##### Upsert state reduction
```

Comment:
This is a bit of a placeholder--need to work on this design with @moulimukherjee but just wanted to put up something.

Reply:
We'll also need to think about restarts, i.e. upon restart of a cluster, how do we populate this upsert state so that it still has the last 24 hours of data. Currently in upsert we are either storing it in-memory or for spill to disk with rocksdb (which is worker local), both of which we rehydrate back from persist upon restart.

Reply:
OOC does persist support filter by timestamps? So something like `now() - 1 day`?